allow changing timestamp cache size

clockfort opened the pull request:
also a little bit of drive-by cleanup to some still-not-great test classes

**Goals (and why)**:
There is an internal service whose basic premise is to catalogue small pieces of metadata, and build up large views of this data from these many individual smaller pieces/transactions.

As such, this service ends up commonly reading a plethora of historical commit timestamps across all points in time, and has a large active working set of read-in historical transactions.

Because the working set trends to higher than the current 1M timestamp cache size, I'm told by @tpetracca that this service has poor timestamp cache hit rate which in turn increases query latency from roundtrips on cache miss.

This is an easy stop-gap measure; the harder long term real fix is to implement transaction table sweeping, which I am currently looking into the requirements for, and have already found at least one issue with that needs to be fixed internally before I can start adding things to atlas. (notably, there is an internal product that has its own stack of nested KVSs and is accidentally bypassing sweep knowing about some of its tables)

**Priority (whenever / two weeks / yesterday)**:
whenever

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/palantir/atlasdb/2496)
<!-- Reviewable:end -->

tpetracca commented on atlasdb-client/src/main/java/com/palantir/atlasdb/AtlasDbConstants.java:
```diff
@@ -115,6 +115,11 @@ private AtlasDbConstants() {
 
     public static final int DEFAULT_STREAM_IN_MEMORY_THRESHOLD = 4 * 1024 * 1024;
 
+
```
nit: too many new lines

tpetracca commented on atlasdb-config/src/main/java/com/palantir/atlasdb/config/AtlasDbConfig.java:
```diff
@@ -254,6 +254,20 @@ public int getDefaultLockTimeoutSeconds() {
+     * and has read patterns that deal with a very large working set of existing transactions.
+     */
+    @Value.Default
+    public long getTimestampCacheSize() {
```
Should this go in `AtlasDbRuntimeConfig` instead so you can resize it to larger/smaller live?

clockfort replied:
It wouldn't be easy to make this live changeable; TimestampCache can't know what an AtlasRuntimeConfig is because of circular dependency problems, and you only initialize the parent SnapshotTransactionManager once.

tpetracca replied:
:/ I feel like we need to address the underlying issue. This makes it public api and yet another thing to break when we want the config / TMs stuff to not be such a mess.

tpetracca commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/memory/InMemoryAtlasDbFactory.java:
```diff
@@ -70,9 +71,16 @@
      */
     @Deprecated
-    private static final int DEFAULT_MAX_CONCURRENT_RANGES = 64;
+    private static final int DEFAULT_MAX_CONCURRENT_RANGES = AtlasDbConstants.DEFAULT_CONCURRENT_RANGES_PER_QUERY;
```
Feels weird that this value should change (you set DEFAULT_CONCURRENT_RANGES_PER_QUERY to 32 not 64 above)

tpetracca replied:
Actually I don't care :shrug:

clockfort replied:
This was on purpose fixing all of the tests, they were previously random values of 4, 16, 64, and 32 range concurrency, so I just made them all just one value.

tpetracca commented on atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionTest.java:
```diff
@@ -87,13 +87,13 @@
 @SuppressWarnings({"checkstyle:all","DefaultCharset"}) // TODO(someonebored): clean this horrible test class up!
```
totally unrelated lol @ this line

schlosna commented on atlasdb-client/src/main/java/com/palantir/atlasdb/cache/TimestampCache.java:
```diff
@@ -15,26 +15,34 @@
 public class TimestampCache {
 
-    private final Cache<Long, Long> startToCommitTimestampCache;
+    private Cache<Long, Long> startToCommitTimestampCache;
```
Does this need to be volatile?

schlosna commented on atlasdb-client/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractTransactionManager.java:
```diff
@@ -42,7 +42,7 @@
 
     public static final Logger log = LoggerFactory.getLogger(AbstractTransactionManager.class);
-    protected final TimestampCache timestampValidationReadCache = TimestampCache.create();
+    protected final TimestampCache timestampValidationReadCache = new TimestampCache();
```
Should we create with initial desired size in constructor?

clockfort replied:
it's difficult, working around some architectural problems because things in AbstractTransactionManager can't very easily get anything from an atlas config (notably, some ways to get into this code don't even have an atlas config)

clockfort commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java:
```diff
@@ -85,7 +85,7 @@ public void registerClosingCallback(Runnable closingCallback) {
      */
     protected SerializableTransactionManager() {
-        this(null, null, null, null, null, null, null, null, null, 1);
+        this(null, null, null, null, null, null, null, null, null, 1, AtlasDbConstants.DEFAULT_TIMESTAMP_CACHE_SIZE);
```
all of the transaction manager code is the worst, in case this line doesn't make things clear :-/

gmaretic replied:
There's no reason to maintain a large cache for the dummy transaction manager. We can just do `this(null, null, null, null, null, null, null, null, null, 1, 1);`?

gmaretic commented on atlasdb-client/src/main/java/com/palantir/atlasdb/cache/TimestampCache.java:
```diff
@@ -15,26 +15,34 @@
-    private final Cache<Long, Long> startToCommitTimestampCache;
+    private Cache<Long, Long> startToCommitTimestampCache;
 
-    public static TimestampCache create() {
```
This is a (admittedly minimal) devbreak, so should probably mention it in the release notes.

gmaretic commented on atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransactionManager.java:
```diff
@@ -85,7 +85,7 @@ public void registerClosingCallback(Runnable closingCallback) {
+        this(null, null, null, null, null, null, null, null, null, 1, AtlasDbConstants.DEFAULT_TIMESTAMP_CACHE_SIZE);
     }
 
     public static SerializableTransactionManager create(KeyValueService keyValueService,
```
I like not bloating with additional constructors, but again this is a minor dev break that we should mention.

clockfort commented:
rebased onto develop:head,
dealt with the 20 content merge conflicts jeff left as a present for me,
added an additional commit to address gmaretic comments
